Add featured image support to Social Web reader

pfefferle opened the pull request:
## Summary
Adds featured image display capability to the ActivityPub Social Web reader feed view.

## Changes
- **Backend**: Store first image from ActivityPub attachments as `_activitypub_featured_image_url` post meta
- **REST API**: Register `featured_image` field on `ap_post` post type
- **Frontend**: New `featuredImageField` component for DataViews
- **Types**: Add `featured_image` to `FeedPost` interface and `useFeed` hook defaults
- **Styling**: Flex-based layout with image positioned on right side via CSS

## Implementation Details
- Featured images are cached locally during attachment import process
- Stored in post meta for efficient retrieval (no filesystem scanning)
- Meta is cleaned up on post updates
- Users can enable/disable the image field via DataViews properties panel
- Image appears as first field in field order, styled to display on the right

## Testing
1. Follow an ActivityPub actor that posts images
2. View the Social Web feed page
3. Enable the "Image" field in DataViews properties
4. Verify images display on the right side of feed items

## Checklist
- [x] Code follows WordPress coding standards
- [x] Changes are properly documented
- [x] TypeScript types are updated
- [x] Build succeeds without errors

pfefferle committed:
Add featured image support to Social Web reader

Implements featured image display in the ActivityPub reader feed view:

- Store first image from ActivityPub attachments as post meta
- Register REST API field for featured_image on ap_post post type
- Create featured image field component for DataViews
- Add featured_image to feed query and TypeScript types
- Position image on the right side using flex layout
- Clean up meta on post update

The featured image is cached locally during attachment import and
stored in post meta for efficient retrieval. Users can enable/disable
the image field via DataViews properties panel.

obenland commented:
Can you help me understand the benefit of designating a featured image?

pfefferle commented:
that was a feature that was requested at the meetup! to have at least one image show up in the stream view.

/cc @jeherve

obenland commented:
Have at least one picture show up when excerpt is selected?

pfefferle commented:
yes

pfefferle committed:
Merge add/reader into add/reader-featured-image

pfefferle committed:
Add phpcs ignore for unused variable in attachments

Added a phpcs:ignore comment to suppress VariableAnalysis warnings for the unused $old_url variable in the inline mappings loop, clarifying that only the new URL is needed.

pfefferle committed:
Display featured image within excerpt field

Integrate featured image display into the excerpt field instead of showing it as a separate column. The image now appears below the excerpt text with responsive sizing constraints.

Changes:
- Move featured image rendering into excerpt field component
- Add SCSS styling for excerpt images (300px max dimensions, auto width/height to maintain aspect ratio)
- Remove standalone featured image field from feed stage
- Remove featured image export from fields index
- Image displays with 1em top margin and 4px border radius
- Update built assets to reflect source changes

pfefferle commented:
Updated the featured image implementation:

- Featured image is now displayed within the excerpt field instead of as a separate column
- Image appears below the excerpt text with proper spacing
- Image sizing: 300px max width/height, maintains aspect ratio
- Only shows in excerpt view, not in content view
- Responsive design with auto width/height

pfefferle committed:
Fix accessibility: Decode HTML entities in image alt text

The featured image in the excerpt field now properly decodes HTML entities
in the post title when used as alt text. This ensures that titles containing
entities like &amp;, &quot;, etc. are properly displayed to screen readers
and other assistive technologies.

pfefferle commented:
## Copilot Feedback Review

I've reviewed all 7 Copilot comments on this PR. Here's the status:

### ✅ Fixed

**HTML entities in alt text (excerpt/index.tsx:36)**
- Valid accessibility issue
- Fixed by adding `decodeEntities()` to the image alt attribute
- Ensures titles with entities like `&amp;`, `&quot;` are properly displayed to screen readers

### ❌ Not Applicable

The remaining 6 comments reference the standalone `featured-image` field component, which has been removed from this implementation. The featured image is now integrated directly into the excerpt field instead of being a separate column. The referenced code is no longer used or exported.

**Specific comments not applicable:**
1. Browser support for `:has()` pseudo-class - references unused component
2. Duplication between featured-image and excerpt components - excerpt now handles both
3. Unused `register()` call return value - standard REST API pattern, not an issue
4. Featured image width constraints - component no longer in use
5. REST API response structure - standard WordPress pattern
6. Additional duplication comment - resolved by integration approach

The new implementation displays the featured image within the excerpt field, maintaining aspect ratio with a 300px max dimension constraint. This provides a cleaner UI with the image positioned below the excerpt text.

obenland commented:
@jeherve @pfefferle Can you help me understand the need for this?

> to have at least one image show up in the stream view.

Why wouldn't I use the content view in that case?

The more I use the app myself, the more I wonder if we should remove the excerpt view all-together. It looks pretty bad for Note content.

pfefferle commented:
I am quite the opposite. some full blog post breaks the readability a lot!

obenland commented:
Sure, but I wouldn't expect those to show up as Notes? And if they do that's probably something we should address with the platform they use.

Maybe we should add that short form/long form switch sooner rather than later.

pfefferle commented:
Are we talking about the same thing? This is about having the excerpt in the main view not?

What has this todo with the object-type filter?

pfefferle commented:
> Maybe we should add that short form/long form switch sooner rather than later.

what is that?

pfefferle committed:
Merge add/reader branch into add/reader-featured-image

pfefferle committed:
Move featured image support from excerpt to content field

Since the excerpt field was removed in the add/reader branch, this change
migrates the featured image functionality into the content field instead.

Changes:
- Add featured image rendering to content field for both Notes (full content) and Articles (excerpt)
- Decode HTML entities in alt text for accessibility
- Add image styles for both content and excerpt contexts
- Images display with 300px max dimensions and 4px border radius
- Images have 1em top margin for proper spacing

pfefferle committed:
Remove excerpt field directory

Clean up the excerpt field directory that was left over from
the migration to content field. The excerpt field was removed
in the add/reader branch and its functionality has been merged
into the content field.

obenland commented on includes/collection/class-posts.php:
```diff
@@ -325,7 +326,22 @@ private static function maybe_import_attachments( $activity_object, $post_id ) {
-			Attachments::import_post_files( $activity_object['attachment'], $post_id );
+			$files = Attachments::import_post_files( $activity_object['attachment'], $post_id );
+
+			// Save the featured image URL as post meta for efficient retrieval.
+			if ( ! empty( $files ) ) {
+				foreach ( $files as $file ) {
+					// Check if this is an image file.
+					if ( isset( $file['mime_type'] ) && 0 === strpos( $file['mime_type'], 'image/' ) ) {
+						\update_post_meta(
+							$post_id,
+							'_activitypub_featured_image_url',
+							$file['url']
+						);
+						break; // Only save the first image.
+					}
+				}
+			}
```
This should happen as part of `Attachments::import_post_files()`

obenland commented on includes/class-post-types.php:
```diff
@@ -406,6 +406,17 @@ public static function register_post_post_type() {
 			)
 		);
+
+		\register_post_meta(
+			Posts::POST_TYPE,
+			'_activitypub_featured_image_url',
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'description'       => 'URL of the featured image from ActivityPub attachments.',
+				'sanitize_callback' => 'sanitize_url',
+			)
+		);
```
Instead of using custom storage we should add `'thumbnail'` support to the custom post type registration and re-use Core APIs.

pfefferle replied:
but doesn't that need an attachment?

obenland commented on src/social-web/components/fields/content/index.tsx:
```diff
@@ -47,13 +48,37 @@ export const contentField: Field< FeedPost > = {
 						className="activitypub-feed-content"
 						dangerouslySetInnerHTML={ { __html: content || '<p>\u00A0</p>' } }
 					/>
+					{ hasFeaturedImage && (
+						<img
+							src={ item.featured_image }
+							alt={ decodeEntities( item.title?.rendered || __( 'Post image', 'activitypub' ) ) }
+							className="activitypub-feed-content__image"
+							loading="lazy"
+						/>
+					) }
```
We shouldn't add a featured image to Notes, it'll likely duplicate an image that gets displayed as part of the content.

pfefferle replied:
Yes, that was an unfinalized port from the excerpt/content system to the new "magic" content system.

obenland commented on src/social-web/components/fields/content/index.tsx:
```diff
@@ -47,13 +48,37 @@ export const contentField: Field< FeedPost > = {
+				{ hasFeaturedImage && (
+					<img
+						src={ item.featured_image }
+						alt={ decodeEntities( item.title?.rendered || __( 'Post image', 'activitypub' ) ) }
```
```suggestion
						alt={ decodeEntities( item.title?.rendered || __( 'Post Thumbnail', 'activitypub' ) ) }
```

obenland commented on src/social-web/components/fields/featured-image/index.tsx:
```diff
@@ -0,0 +1,31 @@
+			return null;
+		}
+
+		const altText = item.title?.rendered || __( 'Post image', 'activitypub' );
```
```suggestion
		const altText = item.title?.rendered || __( 'Post Thumbnail', 'activitypub' );
```

obenland reviewed (commented):
I still don't think this is needed. If you feel strongly about having it we should not process them and display them for Note types, though.

pfefferle committed:
Update src/social-web/components/fields/content/index.tsx

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update src/social-web/components/fields/featured-image/index.tsx

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Merge branch 'add/reader' into add/reader-featured-image
